intercom/intercom-php
# Abstract the HTTP client using HTTPPlug

## Why?

As raised in https://github.com/intercom/intercom-php/issues/255, the restriction to use Guzzle6 is a blocker for some users, especially when using other versions of the library.

## How?

[HTTPPlug](http://httplug.io/) allows us to abstract our HTTP Client implementation, and allow the SDK users to decide which client to use. It also provides enough tooling to discover already existing clients, so unless there's an especial need, this SDK will work out of the box. If a more specific implementation is required, you can still pass your custom client using the `setClient` method.

## How to test this?

1. Update your composer.json file, so you set the version `dev-ga/http-plug`.
1. See the upgrade notes at the bottom of this PR

## What's the state of this PR?

✅ This PR has been tested and it is ready to review

## To-do

- [x] Allow HTTP Plug 1.0 and 2.0
- [x] Write release notes

Draft release notes:

> This version introduces [HTTPPlug](https://github.com/php-http/httplug) as the HTTP Client. HTTPPlug allows this library to use any HTTP Library as long as there is a HTTPPlug adaptor for it. All major libraries are available (Guzzle5/6, CURL, Buzz and [many more](http://docs.php-http.org/en/latest/clients.html)).
> 
> Upgrade instructions:
> 1. Check if any change from the list below affects your code and update it if required.
> 1. Choose a [client implementation](http://docs.php-http.org/en/latest/clients.html) and include it to your dependencies (ie `composer require php-http/curl-client`). We recommend that you choose the adapter for the library you already use in your project. If you don't use any, `php-http/curl-client` is the most lightweight.
> 1. Upgrade your `intercom/intercom-php` dependency: `composer update intercom/intercom-php`.
> 
> Changes in this version:
> * `IntercomClient` constructor third parameter now only accepts a key-value array of request headers, that will be included in every request, eg: `new IntercomClient('token', null, ['Custom-Header' => 'value']);`. If you were passing any other options to the client (apart from headers), you will need to instantiate your own client and pass it using the `setHttpClient` method. Example:
> 
>   ```php
>   $httpClient = new Http\Adapter\Guzzle6\Client(
>       new GuzzleHttp\Client(['connect_timeout' => 5]);
>   );
>   $client = new IntercomClient('token');
>   $client->setHttpClient($httpClient); 
>   ```
> 
> * `IntercomClient` method `setClient` has been renamed to `setHttpClient`. Its first argument must be a `Psr\Http\Client\ClientInterface`. If you were using this method before upgrading, you can use the Guzzle6 adapter like in the example above. For example, `$client->setClient($guzzleClient)` would need to be changed to `$client->setHttpClient(new Http\Adapter\Guzzle6\Client($guzzleClient))`. 
> * `IntercomClient` no longer exposes the methods `getGuzzleRequestOptions` and `getAuth`.
> * `IntercomClient` now provides the methods `setRequestFactory` and `setUriFactory` so you can customize the generation of requests and URIs if needed.
> * From now on, all the exceptions thrown by the SDK will implement `Http\Client\Exception`. See the different exceptions that can be thrown [in the HTTPPlug documentation](http://docs.php-http.org/en/latest/httplug/exceptions.html)
